pkg/kinflate/resource/resource.go

```diff
@@ -16,10 +16,46 @@ limitations under the License.
 
 package resource
 
-import "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+import (
+	"encoding/json"
 
-// Examples: deployment, service, configmap, secret, etc.
-type Resource interface {
-	Name() string
-	Data() unstructured.Unstructured
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/kubectl/pkg/kinflate/types"
+)
+
+// Resource represents a Kubernetes Resource Object for ex. Deployment, Server
+// ConfigMap etc.
+type Resource struct {
+	Data *unstructured.Unstructured
+
+	// OrigName is the name of this resource at the time of of loading it from a
+	// source. Actual name may be differ because we add suffix for configmaps
+	// or name-prefix.
+	OrigName string
```

Comment:
I really don't want to block, but this still seem quite wrong. That is an un-pumped special case in a general structure.

Can we append the hash to the config map as a transformation? Then we give a chance for whoever created this resource to save the name. e.g.

```
r, _ := NewFromConfigMap(cm) // Create the resource from manifest.ConfigMap
oldName := r.GVKN().Name // Save its name
r := AppendHashName(r) // Append the hash
newName := r.GVKN().Name // New name
// s/oldName/newName/ where needed.
```

WDYT?

Reply:
Considering the usage of this `OrigName` for different resources(resource, configmap, secret, patch),  it can be used to change the name from the oldName to newName. Currently only configmap and secret need to have newName, which is with hash suffix. If we don't have this OrigName, we can still append that suffix.

Reply:
@apelisse @Liujingfang1  I like the idea of handling suffix as a transformer. What that means is: we can combine all the name transformations like name-prefix/hash(only for configmap) in to one transformer. I am liking it even more now. @mengqiy do you see any problem with that ?

Reply:
Removed origName for now. We will try to handle it in the name transformation. If that doesn't work, then we will add it back.